Blog post for ingress-nginx stabilization discussion (#28182)
kubernetes/website: content/en/blog/_posts/2021-06-21-update-with-ingress-nginx.md

---
layout: blog
title: 'Updates to Ingress Nginx with network v1'
date: 2021-06-21
slug: update-with-ingress-nginx
---

> **Authors:** James Strong, Richardo Katz 

Reviewer: ```suggestion
**Authors:** James Strong, Ricardo Katz
```
?

With all Kubernetes APIs, there is a process to creating, maintaining, and

> ultimately deprecating them once they become GA. The networking.k8s.io API is no

Reviewer: ```suggestion
ultimately deprecating them once they become GA. The networking.k8s.io API group is no
```

> different. With the release of Kubernetes 1.22, The networking.k8s.
> io/v1beta1 API version of IngressClass will no longer be served in 
> Kubernetes v1.22. The extensions/v1beta1 and networking.k8s.io/v1beta1 API 
> versions of Ingress also will no longer be served in v1.22.  This move has been 

Reviewer: How about:
```suggestion
different. The upcoming Kubernetes 1.22 release will remove several deprecated APIs
that are relevant to networking:

- the `networking.k8s.io/v1beta1` API version of [IngressClass](/docs/concepts/services-networking/ingress/#ingress-class)
- all beta versions of [Ingress](/docs/concepts/services-networking/ingress/): `extensions/v1beta1` and `networking.k8s.io/v1beta1`

On a v1.22 Kubernetes cluster, you'll be able to access Ingress and IngressClass
objects through the stable (v1) APIs, but access via their beta APIs won't be possible.
This change has been in
 ```
?

in discussion since
[2017](https://github.com/kubernetes/kubernetes/issues/43214),
[2019](https://kubernetes.io/blog/2019/07/18/api-deprecations-in-1-16/) with 
1.16 Kubernetes API deprecations,and most recently in 
KEP-1453: 
[Graduate Ingress API to GA](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1453-ingress-api#122).

> During our community meetings, the group has decided to continue 

Reviewer: Is this correct:
```suggestion
During community meetings, the networking Special Interest Group has decided to continue 
```
?

> supporting Kubernetes versions older than 1.22 with Ingress version 0.47.0. 

Reviewer: There is no v0.47.0 version of the Ingress API. Did you mean Ingress-NGINX instead?

Author: yes, updated to clarify that point

Support for Ingress-NGINX will continue for six months after Kubernetes 1.22 
is released. Any additional bug fixes and CVEs for Ingress-NGINX will be 
addressed on a need-by-need basis.

Ingress-NGINX will have separate branches and releases of Ingress-NGINX to 
support this model, mirroring the Kubernetes project process. Future 
releases of the Ingress-NGINX project will track and support the latest 
versions of Kubernetes.

{{< table caption="Ingress NGINX supported version with Kubernetes Versions" >}}
Kubernetes Version  | Ingress-NGINX version | Support Notes
:-------------------|:----------------------|:------------
`1.22`              | TBD                   | On Going
`1.21`              | `v0.47.0`             | On Going but only CVE and crashes
`1.20`              | `v0.47.0`             | On Going but only CVE and crashes
`1.19`              | `v0.47.0`             | Support will drop 6 months after 1.22 release
{{< /table >}}    

Because of the updates in Kubernetes 1.22, **v0.47.0** will not work with 
Kubernetes 1.22. 

> # What do I need to do

Reviewer: ```suggestion
# What you need to do
```

(“I” doesn't make sense for an article with two authors, and may especially confuse non-native speakers of English).

The team is currently in the process of upgrading ingress-nginx to support 
the v1 migration, you can track the progress 
[here](https://github.com/kubernetes/ingress-nginx/pull/7156).  

> Feature requests are being frozen until the v1 update is completed. 

Reviewer: ```suggestion
We're not making feature improvements to `ingress-nginx` until after the support for
Ingress v1 is complete.
```

In the meantime to ensure no compatibility issues: 

> * Update to the latest version of Ingress-NGINX, currently 

Reviewer: punctuation nit:
```suggestion
* Update to the latest version of Ingress-NGINX; currently
```
(or put the current version details inside parens)

  [v0.47.0](https://github.com/kubernetes/ingress-nginx/releases/tag/controller-v0.47.0) 
* After Kubernetes 1.22 is released, ensure you are using the latest version of 

>   Ingress-NGINX that supports v1 networking api.

Reviewer: ```suggestion
  Ingress-NGINX that supports the stable APIs for Ingress and IngressClass.
```

The community’s feedback and support in this effort is welcome. The

> Ingress-Nginx Sub-project regularly hold community meetings where we discuss

Reviewer: ```suggestion
Ingress-NGINX Sub-project regularly holds community meetings where we discuss
```
?

> this and other issues facing the project, meeting notes and details are
> [here](https://github.com/kubernetes/community/tree/master/sig-network)

Reviewer: ```suggestion
this and other issues facing the project. For more information on the sub-project, 
please see [SIG Network](https://github.com/kubernetes/community/tree/master/sig-network).
```

---
layout: blog

> title: 'Updates to Ingress Nginx with network v1'

Reviewer: Maybe:
```suggestion
title: 'Updating NGINX-Ingress to use the stable Ingress API'
```
?

date: 2021-06-21
slug: update-with-ingress-nginx
---

**Authors:** James Strong, Ricardo Katz

With all Kubernetes APIs, there is a process to creating, maintaining, and
ultimately deprecating them once they become GA. The networking.k8s.io API group is no
different. The upcoming Kubernetes 1.22 release will remove several deprecated APIs
that are relevant to networking:

- the `networking.k8s.io/v1beta1` API version of [IngressClass](/docs/concepts/services-networking/ingress/#ingress-class)
- all beta versions of [Ingress](/docs/concepts/services-networking/ingress/): `extensions/v1beta1` and `networking.k8s.io/v1beta1`

On a v1.22 Kubernetes cluster, you'll be able to access Ingress and IngressClass
objects through the stable (v1) APIs, but access via their beta APIs won't be possible.
This change has been in
in discussion since
[2017](https://github.com/kubernetes/kubernetes/issues/43214),
[2019](https://kubernetes.io/blog/2019/07/18/api-deprecations-in-1-16/) with 

> 1.16 Kubernetes API deprecations,and most recently in 

Reviewer: nit:
```suggestion
1.16 Kubernetes API deprecations, and most recently in
```

KEP-1453: 
[Graduate Ingress API to GA](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1453-ingress-api#122).

During community meetings, the networking Special Interest Group has decided to continue 
supporting Kubernetes versions older than 1.22 with Ingress-NGINX version 0.47.0. 
Support for Ingress-NGINX will continue for six months after Kubernetes 1.22 
is released. Any additional bug fixes and CVEs for Ingress-NGINX will be 
addressed on a need-by-need basis.

Ingress-NGINX will have separate branches and releases of Ingress-NGINX to 
support this model, mirroring the Kubernetes project process. Future 
releases of the Ingress-NGINX project will track and support the latest 
versions of Kubernetes.

{{< table caption="Ingress NGINX supported version with Kubernetes Versions" >}}

> Kubernetes Version  | Ingress-NGINX version | Support Notes

Reviewer: ```suggestion
Kubernetes version  | Ingress-NGINX version | Notes
```

:-------------------|:----------------------|:------------

> `1.22`              | TBD                   | On Going
> `1.21`              | `v0.47.0`             | On Going but only CVE and crashes
> `1.20`              | `v0.47.0`             | On Going but only CVE and crashes
> `1.19`              | `v0.47.0`             | Support will drop 6 months after 1.22 release

Reviewer: Is this accurate?
```suggestion
v1.22              | ≥ v.0.47.0     | New features, plus bug fixes.
v1.21              | v0.47.x        | Bugfixes only, and just for security issues or crashes. No end-of-support date announced.
v1.20              | v0.47.x        | Bugfixes only, and just  for security issues or crashes. No end-of-support date announced.
v1.19              | v0.47.x        | Bugfixes only, and just  for security issues or crashes. Fixes only provided until 6 months after Kubernetes v1.22.0 is released.
```

I changed `v0.47.0` to `v0.47.x` because I assume there would be a new patch release, rather than yanking `v0.47.0` and replacing it with different code.

> ---
> layout: blog
> title: 'Updating NGINX-Ingress to use the stable Ingress API'
> date: 2021-06-21
> slug: update-with-ingress-nginx
> ---

Reviewer: How about we get this merged _as a draft_?
Then, when we want to publish it, it'll be a small diff. We can also accept PRs to keep the draft current during the period before we can get that published.

```suggestion
---
draft: true
layout: blog
title: 'Updating NGINX-Ingress to use the stable Ingress API'
date: 2021-06-21
slug: update-with-ingress-nginx
---
```

Author: updated.
